src/Features/LanguageServer/Protocol/Handler/Completion/CompletionHandler.cs

```diff
@@ -139,13 +168,36 @@ bool IsValidTriggerCharacterForDocument(Document document, char triggerCharacter
                 CompletionItem item,
                 long? resultId,
                 CompletionTrigger completionTrigger,
-                Dictionary<ImmutableArray<CharacterSetModificationRule>, ImmutableArray<string>> commitCharacterRulesCache) where TCompletionItem : LSP.CompletionItem, new()
+                Dictionary<ImmutableArray<CharacterSetModificationRule>, ImmutableArray<string>> commitCharacterRulesCache,
+                bool returnTextEdits,
+                SourceText? documentText = null,
+                TextSpan? defaultSpan = null,
+                LSP.Range? defaultRange = null) where TCompletionItem : LSP.CompletionItem, new()
             {
                 var completeDisplayText = item.DisplayTextPrefix + item.DisplayText + item.DisplayTextSuffix;
+                var insertText = item.Properties.ContainsKey("InsertionText") ? item.Properties["InsertionText"] : completeDisplayText;
```

Comment:
oooh, this is really sketchy.  'InsertionText' is not a supported part of our API.  It's a literal internal impl detail.  COmpletion providers are absolutely free to not use it, or name it something else entirely.

Reply:
i actually don't see how this can work at all.  FOr example the regex completion provider does this:

```c#
private const string NewTextKey = nameof(NewTextKey);
...
        public override Task<CompletionChange> GetChangeAsync(Document document, CompletionItem item, char? commitKey, CancellationToken cancellationToken)
        {
            // These values have always been added by us.
            var startString = item.Properties[StartKey];
            var lengthString = item.Properties[LengthKey];
            var newText = item.Properties[NewTextKey];

            // This value is optionally added in some cases and may not always be there.
            item.Properties.TryGetValue(NewPositionKey, out var newPositionString);

            return Task.FromResult(CompletionChange.Create(
                new TextChange(new TextSpan(int.Parse(startString), int.Parse(lengthString)), newText),
                newPositionString == null ? (int?)null : int.Parse(newPositionString)));
        }
```

This is needed by regex so you can select something like `\A` but so we insert `\\A`.  i think we may need to rethink this approach.

Reply:
Thanks for the catch! In the latest update pushed to the PR, I changed the logic so we now call `GetChangeAsync` directly. This results in us being able to avoid the `InsertionText` hack, and we now behave correctly in this situation (i.e. we insert `\A` instead of ```\\A```.
The fix is behind a feature flag due to TextEdit perf concerns, but hopefully we can remove the feature flag eventually.